src/impl/apple/keychain_helper.cpp

```diff
@@ -50,66 +51,96 @@ CFPtr<CFStringRef> convert_string(const std::string& string)
-    const std::string account = "metadata";
-
     auto search_dictionary = build_search_dictionary(account, service, none);
     CFDataRef retained_key_data;
     if (OSStatus status = SecItemCopyMatching(search_dictionary.get(), (CFTypeRef *)&retained_key_data)) {
-        if (status != errSecItemNotFound) {
+        if (status != errSecItemNotFound)
             throw KeychainAccessException(status);
-        }
 
-        // Key was not found. Generate a new key, store it, and return it.
-        std::vector<char> key(key_size);
-        arc4random_buf(key.data(), key_size);
-        auto key_data = adoptCF(CFDataCreate(nullptr, reinterpret_cast<const UInt8 *>(key.data()), key_size));
-        if (!key_data) {
-            throw std::bad_alloc();
-        }
-
-        CFDictionaryAddValue(search_dictionary.get(), kSecValueData, key_data.get());
-        if (OSStatus status = SecItemAdd(search_dictionary.get(), nullptr)) {
-            throw KeychainAccessException(status);
-        }
-
-        return key;
+        // Key was not found.
+        return none;
     }
-    CFPtr<CFDataRef> key_data = adoptCF(retained_key_data);
 
     // Key was previously stored. Extract it.
-    if (key_size != CFDataGetLength(key_data.get())) {
+    CFPtr<CFDataRef> key_data = adoptCF(retained_key_data);
+    if (key_size != CFDataGetLength(key_data.get()))
         throw std::runtime_error("Password stored in keychain was not expected size.");
-    }
 
     auto key_bytes = reinterpret_cast<const char *>(CFDataGetBytePtr(key_data.get()));
     return std::vector<char>(key_bytes, key_bytes + key_size);
 }
 
+void set_key(const std::vector<char>& key, CFStringRef account, CFStringRef service)
+{
+    auto search_dictionary = build_search_dictionary(account, service, none);
+    auto key_data = adoptCF(CFDataCreate(nullptr, reinterpret_cast<const UInt8 *>(key.data()), key_size));
+    if (!key_data)
+        throw std::bad_alloc();
+
+    CFDictionaryAddValue(search_dictionary.get(), kSecValueData, key_data.get());
+    if (OSStatus status = SecItemAdd(search_dictionary.get(), nullptr))
+        throw KeychainAccessException(status);
 }
+
+}   // anonymous namespace
+
+std::vector<char> metadata_realm_encryption_key(bool check_legacy_service)
+{
+    CFStringRef account = CFSTR("metadata");
+    CFStringRef default_service = CFSTR("io.realm.sync.keychain");
+
+    CFPtr<CFStringRef> managed_service;
+    if (auto bundle_id = retainCF(CFBundleGetIdentifier(CFBundleGetMainBundle()))) {
+        managed_service = adoptCF(CFStringCreateWithFormat(NULL, NULL,
+                                                           CFSTR("%@ - Realm Sync Metadata Key"),
+                                                           bundle_id.get()));
+    } else {
+        managed_service = retainCF(default_service);
+        check_legacy_service = false;
+    }
+
+    // Try retrieving the key.
+    CFStringRef service = managed_service.get();
```

Comment:
I'd be inclined to ditch this local and just add the `.get()` calls in the handful of places it is necessary. It's effectively free to call, and would avoid the need for the slightly odd `managed_` prefix on the variable.